### ISSUE TITLE
Null safety

### ISSUE THREAD

**Issue by noordawod** [vitusortner/floor#375]
Hi again.

Is there a plan to support null safety for the generator sometime in the future?

Thanks.

**Comment by mqus** [comment]
I assume that you're referring to the [null safety of dart](https://medium.com/dartlang/announcing-sound-null-safety-defd2216a6f3) and how we will interact with non-nullable types in the code analysis and generation. And (speaking for me) there absolutely is one. 

The sqlparser integration(still WIP but complete feature-wise) already supports type derivation of (non-)null query parameters and type checking of (non-)nullable returns. Implementing this should mostly consist of more tests and maybe ten lines of code in the actual generator. Columns of entities and views already have a nullability assigned to it through the `@ColumnInfo` annotation. I'm not sure on how to proceed there yet, but I think we should use some combination of type/deprecation warnings and type checking errors while slowly phasing out the `nullable` parameter. This also depends on how optional the non-nullable types are (this is something I haven't checked out yet).

Let's track this feature here.

I've written a lot of Java and a bit of Rust before I started dabbling in Dart and the decision to have nullable types (even primitives like booleans!) seemed really backwards to me, so I'm very happy they found a solution and I am all for also supporting it in floor.


Btw, internally, vitus already introduced [`@nullable` and `@nonNull`](https://github.com/vitusortner/floor/blob/develop/floor_generator/lib/misc/annotations.dart) for annotating (non-)null returns of internal functions, just to get us ready for also using nonnullable types inside the floor codebase. This won't matter for users of floor but will probably ease our pain when migrating a lot more, even if the annotations were only cosmetically until now.

**Comment by vitusortner** [comment]
I do agree with everything @mqus mentioned. There is high interest in supporting Dart's non-nullable types!

**Comment by mqus** [comment]
Bookmark with lots of useful details: https://dart.dev/null-safety/understanding-null-safety

**Comment by vitusortner** [comment]
Work on null safety support has been started here https://github.com/vitusortner/floor/tree/nullsafety

**Comment by dkunzler** [comment]
Nice that you are working on it. 
Would you mind publishing your branch as an unstable version to pub because I want to try it out.
If I include the dependency like so:
```
  floor:
    git:
      url: https://github.com/vitusortner/floor
      path: floor
      ref: nullsafety
```
I get the error `Invalid description in the "floor" pubspec on the "floor_annotation" dependency: "../floor_annotation/" is a relative path, but this isn't a local pubspec.` Even if I try to override the floor_annotation dependency it doesn't work.

**Comment by vitusortner** [comment]
Hi @dkunzler! I'll look into creating a nullsafety release. Until then, you can clone the repository and use the project that way. Dependency overrides have been faulty for a while and unluckily don't work in this scenario.

**Comment by vitusortner** [comment]
You can get the first null safe version of floor now on pub https://pub.dev/packages/floor/versions/1.0.0-nullsafety.0 🚀

**Comment by dkunzler** [comment]
Thank you, I appreciate your fast response!

**Comment by marcosmko** [comment]
We got flutter 2.0 the day before yesterday, any forecast on when this package will be updated?

Thanks!
@vitusortner

**Comment by vitusortner** [comment]
We're looking into it right now and we'll create a new null-safe release as soon as possible.

**Comment by vitusortner** [comment]
We just published the 1.0.0-nullsafety.1 release with support for Dart 2.12.0. We unluckily can't go stable as crucial dependencies haven't yet been migrated. If you encounter problems with 1.0.0-nullsafety.1, please raise a new issue on GitHub.

**Comment by vitusortner** [comment]
The stable, null-safe 1.0.0 release has just been published. Thanks to everyone that helped! 🎉
